Replace glibc's error() with own implementation

davidpolverari opened the pull request:
This avoids the dependency on glibc's non-standard `error()` function with our own implementation. This is done to enable usage of other C standard library implementations.

hartwork committed:
full-check.yml: Cover compilation against musl libc

hartwork committed:
src/sha2.h: Replace SHA2_USE_INTTYPES_H by HAVE_INTTYPES_H for musl libc

Symptom was:
> musl-gcc -DHAVE_CONFIG_H -I. -I..    -fgnu89-inline -std=c99 -Wall -Wextra -pedantic -MT md5.o -MD -MP -MF .deps/md5.Tpo -c -o md5.o md5.c
> In file included from hash.h:31,
>                  from md5.h:49,
>                  from md5.c:35:
> sha2.h:76:9: error: unknown type name ‘u_int8_t’
>    76 | typedef u_int8_t  sha2_byte;    /* Exactly 1 byte */
>       |         ^~~~~~~~

hartwork committed:
full-check.yml: Dump config.log on configure failure for debugging

hartwork committed:
full-check.yml: Exclude sanitizers from musl-gcc

davidpolverari committed:
configure.ac: enable assertions by default

This checks whether to enable assertions in the style of assert.h.
Assertions are enabled by default, but can be disabled by invoking
configure with the `--disable-assert` option.

davidpolverari committed:
Add dcfldd_error module

This adds the dcfldd_error module, which implements the dcfldd_error()
function as a replacement for the glibc's error function. This is to
avoid a dependency on a non-standard libc function, making it easier to
build dcfldd with other libc implementations.

davidpolverari committed:
Replace error() invocations by dcfldd_error()

hartwork commented on src/dcfldd_error.c:
```diff
@@ -0,0 +1,61 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+#include "config.h"
```
```suggestion
#if HAVE_CONFIG_H
#  include "config.h"
#endif
```
(e.g. for consistency with at least some other files)

davidpolverari replied:
We are using autotools, so I think we don't need to make it conditional as `config.h` is always there (I think? maybe I'm wrong, going to have a look).

I think other files have those conditional preprocessor directives due to historical reasons. Anyways, it is something we could tackle in other PRs, what do you think?

hartwork replied:
@davidpolverari I'm good with not adding the guard and dropping all the other guards in a dedicated pull request.

hartwork commented on src/dcfldd_error.c:
```diff
@@ -0,0 +1,61 @@
+ *
+ * WARNING: this function is not intended as a full drop-in replacement for
+ * error(). One of the most glaring differences from error() is the usage of
+ * dcfldd's program_name instead of program_invocation_name.
```
I'm starting to wonder if that difference is true and if we could make it more concrete or visual. With…

```console
# git --no-pager grep "program_name = "
src/argmatch.c:  program_name = (char *) argv[0];
src/dcfldd.c:    program_name = argv[0];
src/xstrtol.c:  program_name = argv[0];
```

…we may currently match (rather than deviate from) glibc behavior (related to https://github.com/resurrecting-open-source-projects/dcfldd/pull/34#issuecomment-2416779909). Maybe I'm misunderstanding something?

davidpolverari replied:
I think these files all use `program_name` due to being pulled from coreutils. Their common usage of `program_name`  seems to stem from this fact, and looks unrelated to `error()` usage anyways (there is no mention of `program_name` on `error()`'s manpage).

Besides, apart from `src/dcfldd.c`, the other two set `program_name` because they have a `main()` function for testing purposes, enabled only when `TEST` (in the case of `argmatch.c`) or `TESTING_XSTRTO` (in the case of `xstrtol.c`) is defined.

Do you have any other information about `program_name` being used in `error()`?

hartwork replied:
@davidpolverari let me try to rephrase what I am not yet understanding:
The `dcfldd_error` implementation added here seems to do the same with regard to the text starting the terminal output, yet the comment is talking about a glaring different. I do not understand the different and I do not know what `program_invocation_name` refers to in detail. "dcfldd's program_name" refers to the variable, but the rest is not clear. Does that make sense?

davidpolverari replied:
I see. Although _for dcfldd_ our function does the same as `error()`, it is not 100% compliant to it. This is what is stated in `error(3)` man page:

> The program name printed by error() is the value of the global variable program_invocation_name(3). program_invocation_name initially has the same value as main()'s argv[0]. The value of this variable can be modified to change the output of error()

In our `dcfldd_error()` implementation, we use `program_name` set by `src/dcfldd.c` to get the `argv[0]` contents. Even though the final result is the same, we do not honor `program_invocation_name` anywhere in our code, nor we set it. Thus, we can't claim to be a drop-in replacement to `error()`: if someone sets the global `program_invocation_name` to something, our function will not use it to display the program name. It will use `program_name`, as set by `src/dcfldd.c`'s main function.

Was that the difference you were talking about?

hartwork replied:
@davidpolverari interesting! I think the options forward are:

- a) add a dedicated commit to replace `program_name` by `program_invocation_name` in dcfldd and then the comment can go or
- b) make the comment explain itself more, e.g. by mentioning the man page in the sentence about `program_invocation_name`, saying it's a global variable, or quoting the man page and roughly saying "that's something we don't do". Ideally the comment works well enough for someone who's not one of us two and that has no other context than the file and the comment.

I'd be happy with either of the two. What do you think?

davidpolverari replied:
I was not inclined to tackle a), but after perusing libc's `error()` implementation, I saw this:

```c
#ifdef _LIBC
/* In the GNU C library, there is a predefined variable for this.  */

# define program_name program_invocation_name
```

And then, later on the same file:

```c
/* The calling program should define program_name and set it to the
   name of the executing program.  */
extern char *program_name;
``` 

And on the rest of the code, it used `program_name`:

```c
      fprintf (stderr, "%s: ", program_name);
```

I'm still _more inclined_ not to invest a lot of time making it really an `error()` drop-in replacement, because I don't see it being used that much. In the other hand, if all we needed to add was just that `#define`, I don't see why not. Thus, I will implement that and remove the comment.

davidpolverari replied:
I just pushed https://github.com/resurrecting-open-source-projects/dcfldd/pull/37/commits/f5947afb64f275483c6c6ae2f0a37f0befff4368. Tell me what you think about it.

hartwork replied:
@davidpolverari I think I understand the idea: you want to mimic that glibc feature when compiled against glibc where available and divert use of our own external variable to glibc's external variable. After some experimentation, I see a few issues with this approach and would like to suggest to not go this route and do something else instead, at the end. Bottom up, here's what I found and concluded for myself:

I found man page https://man7.org/linux/man-pages/man3/program_invocation_name.3.html and after playing with example…

```c
#define _GNU_SOURCE

#include <stdio.h>
#include <errno.h>

int main() {
  printf("program_invocation_name = \"%s\"\n",
         program_invocation_name ? program_invocation_name : "NULL!?");
  return 0;
}
```

…I can confirm that with glibc both early `#define _GNU_SOURCE` and `#include <errno.h>` are needed for symbol `program_invocation_name` to available. So if we were to use that symbol, those two things would likely be needed on our side to be robust.

The next thing I checked was define `_LIBC` and I found other (non-trivial) recommendations at https://sourceforge.net/p/predef/wiki/Libraries/#gnu-glibc for a check for glibc.

I believe the two most important questions to ask now are:
- when 100% compatibility with glibc's `error(..)` is still a documented non-goal and not something required for dcfldd, is a complex and "risky" 99% implementation better than a clean and simple 90% implementation? (A clear vote for clean and simple from my side in this context.)
- is different behavior with glibc a feature or bug with regard to dcfld? (My answer would be "a bug" meaning that use of `program_invocation_name` goes against cross-libc compatibility: users on Alpine could get different output than users of Ubuntu.)

With all that said, I'm in clear favor of a comment explaining why we do not support `program_invocation_name` of glibc intentionally. That comment would be the icing on the cake of a very clean re-implementation of `error()` and make a great result in my book.

What do you think?

davidpolverari replied:
I agree. I think we are paying too much attention to a function that is not going to be used outside of dcfldd. A simple comment will do for any contributor wanting to use it when needed.

I will bring back the old comment, expand it a little bit, and remove support for `program_invocation_name`. Suggestions on the comment are welcome.

hartwork replied:
@davidpolverari sounds great, thank you! :+1:

hartwork replied:
@davidpolverari how about?:

```suggestion
 * WARNING: this function is not intended as a full drop-in replacement for
 * error(). One example difference is that glibc's error function reads the
 * program name from a global variable "program_invocation_short_name"
 * (see "man 3 program_invocation_name") whereas the implementation here
 * is using dcfldd's own global variable "program_name", instead.
```

hartwork commented on src/dcfldd_error.c:
```diff
@@ -0,0 +1,61 @@
+#include <stdlib.h>
+#include <string.h>
+
+#include "dcfldd.h"	// to use program_name
```
(could maybe(?) be moved up right after `config.h` but just an idea)

davidpolverari replied:
Hmm, I usually like to include user include files after system ones, although `config.h` is a special case, as `assert.h` will need to know whether `NDEBUG` is defined before being included. That's why I included `config.h` before the system includes, and `dcfldd.h` after. Does that feel right?

hartwork replied:
@davidpolverari nevermind, it's intentional then, alright :+1:

hartwork reviewed (commented):
@davidpolverari looks great! I found three small details:

davidpolverari committed:
dcfldd_error: support program_invocation_name

This makes our error() implementation more glibc-like.

hartwork commented on src/dcfldd_error.c:
```diff
@@ -0,0 +1,66 @@
+/*
+ * dcfldd_error(): a replacement for glibc's error() for usage within dcfldd
+ *
+ * WARNING: this function is not intended as a full drop-in replacement for
+ * error(). One of the most glaring differences from error() is the usage of
+ * dcfldd's program_name instead of program_invocation_name.
```
@davidpolverari how about?:

```suggestion
 * WARNING: this function is not intended as a full drop-in replacement for
 * error(). One example difference is that glibc's error function reads the
 * program name from a global variable "program_invocation_name"
 * (see "man 3 program_invocation_name") whereas the implementation here
 * is using dcfldd's own global variable "program_name", instead.
```

davidpolverari replied:
LGTM. I will use that text, only with some formatting applied, and will remove `program_invocation_name` support.

davidpolverari committed:
dcfldd_error: drop program_invocation_name support

dcfldd_error() is meant to replace only error() invocations within the
dcfldd codebase. Full glibc compliance is not a goal.

Thus, this drops the (incomplete) program_invocation_support, while
better commenting the differences from glibc's error().

davidpolverari commented:
What do you think about it now? Could you test it and see if this solves https://bugs.gentoo.org/713646 for you? Do you think any other changes are needed?

hartwork reviewed (approved):
@davidpolverari looks great to me now! :+1:

davidpolverari commented:
Ok, going to merge it then.

PS: you might want to use `--disable-assert` on the configure step when building from now on. I will put this on the changelog when I release it.

hartwork commented:
> Ok, going to merge it then.

@davidpolverari :+1: 

> PS: you might want to use `--disable-assert` on the configure step when building from now on. I will put this on the changelog when I release it.

I saw the addition but I think I want assertions active in production.

hartwork commented:
@davidpolverari branches `support-musl` and `dcfldd-error` can now be deleted if you like.
